urlfb310a588a_alokhan_LinkStateNetworkSimulation_tgz-pJ8-reso_examples_static_routing_DemoJ8/src/reso/utilities/NetworkBuilder.java

```diff
@@ -0,0 +1,240 @@
+		IPRouter router= createRouter(network, name, new IPAddress[] {}, new EthernetAddress[] {});
+		return router;
+	}
+	
+	private static void loadTopologyRouterLoopback(List<String> tokens, IPRouter router)
+	throws Exception {
+		if (tokens.size() != 2)
+			throw new Exception("Invalid number of parameters in lo interface declaration");
+		IPInterfaceAdapter lo= new IPLoopbackAdapter(router.getIPLayer(), 0);
+		router.getIPLayer().addInterface(lo);
+		IPAddress addr= IPAddress.getByAddress(tokens.get(1));
+		lo.addAddress(addr);
+		lo.setMetric(0);
+		router.getIPLayer().addInterface(lo);
+		router.getIPLayer().addRoute(addr, lo);
+	}
+	
+	private static void loadTopologyRouterEthernet(List<String> tokens, IPRouter router)
+	throws Exception {
+		if (tokens.size() != 2)
+			throw new Exception("Invalid number of parameters in eth interface declaration");
+		EthernetInterface eth= new EthernetInterface(router, Network.generateEthernetAddress());
+		router.addInterface(eth);
+		IPAddress addr= IPAddress.getByAddress(tokens.get(1));
+		IPInterfaceAdapter ip_eth= new IPEthernetAdapter(router.getIPLayer(), eth);
+		ip_eth.addAddress(addr);
+		router.getIPLayer().addInterface(ip_eth);
+	}
+	
+	private static Link<?> loadTopologyLink(List<String> tokens, Network network) throws Exception {
+		if (tokens.size() != 6)
+			throw new Exception("Invalid number of parameters in link declaration");
+		Node n1= network.getNodeByName(tokens.get(1));
+		HardwareInterface<? extends Message> if1= n1.getInterfaceByName(tokens.get(2));
+		Node n2= network.getNodeByName(tokens.get(3));
+		HardwareInterface<? extends Message> if2= n2.getInterfaceByName(tokens.get(4));
+		if (!if1.getType().equals("eth") || !if2.getType().equals("eth"))
+			throw new Exception("cannot connect interfaces [" + n1 + "." + if1 + "," + n2 + "." + if2 + "]");
+		return new Link<EthernetFrame>((EthernetInterface) if1, (EthernetInterface) if2,
+			Double.valueOf(tokens.get(5)));
+	}
+	
+	private static void loadTopologyLinkMetric(List<String> tokens, Link<?> link) throws Exception {
+		if ((tokens.size() < 2) || (tokens.size() > 3))
+			throw new Exception("Invalid number of parameters in link metric declaration");
+		HardwareInterface<? extends Message> if1= link.getHead();
+		IPInterfaceAdapter ip_if1= ((IPHost) if1.getNode()).getIPLayer().getInterfaceByName(if1.getName());
+		HardwareInterface<? extends Message> if2= link.getTail();
+		IPInterfaceAdapter ip_if2= ((IPHost) if2.getNode()).getIPLayer().getInterfaceByName(if2.getName());
+		String token= tokens.get(1);
+		int metric;
+		if (token.equals("?"))
+			metric= Integer.MAX_VALUE;
+		else
+			metric= Integer.valueOf(token);
+		ip_if1.setMetric(metric);
+		
+		if (tokens.size() == 3) {
+			token= tokens.get(2);
+			if (token.equals("?"))
+				metric= Integer.MAX_VALUE;
+			else
+				metric= Integer.valueOf(token);
+		}
+		ip_if2.setMetric(metric);
+	}
+	
+	public static Network loadTopology(String filename, AbstractScheduler scheduler)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.